evaluator/functions.go

```diff
@@ -1319,3 +1346,57 @@ func sleepFn(tok token.Token, args ...object.Object) object.Object {
 
 	return NULL
 }
+
+// source("fileName")
+// aka src(), include(), import(), require()
+var sourceMaxRecursionDepth = 5
+
+func sourceFn(tok token.Token, args ...object.Object) object.Object {
+	// limit source file recursion depth
+	if sourceMaxRecursionDepth <= 0 {
+		return newError(tok, "maximum source file recursion depth exceeded")
+	}
+	// mark this source recursion level
+	sourceMaxRecursionDepth--
+
+	err := validateArgs(tok, "source", args, 1, [][]string{{object.STRING_OBJ}})
+	if err != nil {
+		return err
+	}
+	// load the source file
+	fileName, _ := util.ExpandPath(args[0].Inspect())
+	code, error := ioutil.ReadFile(fileName)
+	if error != nil {
+		// source file path does not exist
+		return newError(tok, "source file not found: %s", fileName)
+	}
+	// parse it
+	l := lexer.New(string(code))
+	p := parser.New(l)
+	program := p.ParseProgram()
+	errors := p.Errors()
+	if len(errors) != 0 {
+		errMsg := fmt.Sprintf("%s", " parser errors:\n")
+		for _, msg := range errors {
+			errMsg += fmt.Sprintf("%s", "\t"+msg+"\n")
+		}
+		return newError(tok, "source file: %s\n%s", fileName, errMsg)
+	}
+	// invoke BeginEval() passing in the sourced program, globalEnv, and our lexer
+	// we save the current global lexer and restore it after we return from BeginEval()
+	savedLexer := lex
+	evaluated := BeginEval(program, globalEnv, l)
+	lex = savedLexer
+	if evaluated != nil {
+		isError := evaluated.Type() == object.ERROR_OBJ
+		if isError {
+			errMsg := evaluated.Inspect()
+			fmt.Printf("%s\n", errMsg)
+			return newError(tok, "found in source file: %s", fileName)
+		}
+	}
+	// restore this source recursion level
+	sourceMaxRecursionDepth++
+
+	return NULL
```

Comment:
One might want to have a sourced file that returns a value. Image you have `date.abs` that looks like:

```
return `date`
```

Then you could do `date = source("date.abs")`. Silly example but just to illustrate that we might want to return what's being evaluated.

Reply:
Done